File: leetcode/cpp/two_sum.cpp
#include "leetcode/cpp/two_sum.hpp"

#include <cstddef>
#include <stdexcept>
#include <unordered_map>

namespace leetcode {

auto two_sum(const std::vector<int>& nums, int target) -> std::array<int, 2> {
  std::unordered_map<int, std::size_t> seen;
  seen.reserve(nums.size());

  for (std::size_t index = 0; index < nums.size(); ++index) {
    const int complement = target - nums[index];
    if (const auto match = seen.find(complement); match != seen.end()) {
      return {static_cast<int>(match->second), static_cast<int>(index)};
    }
    seen.emplace(nums[index], index);
  }

  throw std::invalid_argument{"no two-sum solution exists"};
}

}  // namespace leetcode

File: leetcode/python/two_sum.py
"""LeetCode 1: Two Sum."""


def two_sum(nums: list[int], target: int) -> tuple[int, int]:
    """Return the indices of two values whose sum equals ``target``.

    Time complexity is O(n); auxiliary space complexity is O(n).
    """
    seen: dict[int, int] = {}
    for index, value in enumerate(nums):
        complement = target - value
        if complement in seen:
            return seen[complement], index
        seen[value] = index

    raise ValueError("no two-sum solution exists")
